### ISSUE TITLE
Vision: AI-Powered Optimizer for Automatic Performance & Cost Optimization

### ISSUE THREAD

**Issue by scttfrdmn** [scttfrdmn/cargoship#167]
## Vision Statement
CargoShip learns from upload patterns to automatically optimize performance, cost, and reliability without user intervention.

**"Make CargoShip smart, not just fast"**

## Current State (Rule-Based)
```
Decision Points:
├─ Chunk size: Fixed algorithm (128MB default)
├─ Compression: Extension-based (.txt=compress, .jpg=skip)
├─ Storage tier: User-specified or INTELLIGENT_TIERING
├─ Worker count: CPU count × 2
└─ Retry strategy: Exponential backoff (fixed)

Limitations:
❌ No learning from historical data
❌ No cross-workload optimization  
❌ No predictive cost modeling
❌ No adaptive tuning
```

### ML Models (Ensemble Approach)

**Model A: Compression Predictor**
- Input: File extension, size, entropy, magic bytes
- Output: Predicted compression ratio, recommended algorithm
- Model: XGBoost (gradient boosted trees)
- Target: 90%+ accuracy within 5% of actual ratio

**Model B: Chunk Size Optimizer**
- Input: Dataset size, file distribution, source storage, bandwidth
- Output: Optimal chunk size (32-512MB), predicted time/cost
- Model: Neural Network (3-layer MLP)  
- Training: Reinforcement learning (reward = throughput/cost)

**Model C: Worker Pool Scaler**
- Input: CPU/RAM, network utilization, pipeline depth
- Output: Optimal worker counts, scaling triggers
- Model: Time-series LSTM
- Goal: Maximize throughput within resource constraints

**Model D: Cost Forecaster**  
- Input: Dataset characteristics, compression predictions, AWS pricing
- Output: Upload cost (±5%), storage projections, tier recommendations
- Model: Ensemble (XGBoost + Prophet)
- Target: 95%+ accuracy

**Model E: Failure Predictor**
- Input: Network stability, S3 errors, upload complexity
- Output: Failure probability, retry strategy, alternative region
- Model: Random Forest Classifier
- Goal: Proactive mitigation, not reactive recovery

### User Experience

**Automatic Mode (Default)**
```bash
$ cargoship create upload /data/project --bucket my-bucket

🤖 AI Optimizer analyzing dataset...
📊 Dataset: 50GB, 100k files (mixed types)
🧠 Predicted compression: 65% (confidence: 92%)
⚙️  Optimal chunk size: 96MB (predicted: 8m 23s)
💰 Estimated cost: $0.52 (±$0.03)
🎯 Storage tier: INTELLIGENT_TIERING (lowest TCO)
✨ Worker pool: 8 scanner, 12 archiver, 16 uploader

Continue? [Y/n]
```

**Expert Mode**
```bash
$ cargoship create upload /data --ai-explain

🤖 AI Decisions:
├─ Chunk size: 96MB
│  └─ Reason: Large files (avg 500MB) benefit from bigger chunks
├─ Compression: zstd (level 3)  
│  └─ Reason: 65% predicted ratio, 23% faster than gzip
├─ Workers: 8/12/16
│  └─ Reason: NAS source bottleneck, more archivers help
└─ Storage tier: INTELLIGENT_TIERING
   └─ Reason: Access pattern suggests infrequent retrieval
```

### Privacy & Ethics

**Privacy-Preserving Design**
```
Data Collected (Opt-In Only):
✅ File sizes, types, counts (metadata)
✅ Compression ratios achieved
✅ Transfer rates, resource usage
✅ Cost data, performance metrics

NOT Collected:
❌ File contents
❌ File names/paths  
❌ Directory structure
❌ User credentials
❌ PII

Techniques:
- Differential privacy (ε=1.0)
- Federated learning (local training)
- Anonymized telemetry
- User consent required
- Opt-out anytime
```

**Model Transparency**
- Open source ML models
- Explainable AI (SHAP values)
- Performance benchmarks published
- A/B testing results shared
- Model versioning tracked

## Implementation Roadmap

**Phase 1: Foundation (Q1 2026)**
- Telemetry infrastructure
- Basic ML pipeline (XGBoost)
- Compression ratio predictor
- Cost forecaster

**Phase 2: Advanced Models (Q2 2026)**
- Chunk size optimizer (RL)
- Worker pool scaler (LSTM)
- Failure predictor

**Phase 3: Integration (Q3 2026)**
- --ai-optimize flag
- Online learning pipeline
- Model registry + versioning

**Phase 4: Intelligence (Q4 2026)**
- Federated learning (cross-user)
- Multi-objective optimization
- Automatic A/B testing
- Continuous improvement

## Success Metrics
- 20% faster uploads (vs rule-based)
- 30% cost reduction (vs naive settings)
- 95%+ cost estimate accuracy
- 50% reduction in manual tuning
- 90% user satisfaction

## Research Questions
1. Cold start problem for new users?
2. Concept drift (changing S3 pricing)?
3. Multi-objective optimization (speed vs cost)?
4. Model size (keep <10MB for CLI)?
5. Adversarial robustness (prevent gaming)?

## Related Work
- AWS S3 Intelligent-Tiering (auto cost optimization)
- Google Cloud Autoclass (ML-based tier selection)  
- MinIO (adaptive erasure coding)

## Related
#34 (Best-in-Class S3 Tool)
#106 (Budget management)
#165 (Cost benchmarking)
#166 (Realistic benchmarks)

Proposal: /tmp/ai-optimizer-proposal.md

**Comment by scttfrdmn** [comment]
## Feasibility & decision analysis (needs-decision)

I did a code-grounded feasibility pass on this vision against what CargoShip **actually has today**. Short version: this is **not greenfield** — every one of the 5 proposed models already has a working deterministic/heuristic engine, and two already have in-memory "learning" scaffolds. The real blocker isn't ML; it's that **there's no persisted, cross-upload telemetry corpus to learn from** (with one genuine exception: cost). That reframes the decision.

### What already exists per proposed model

| Model | Today | Where |
|-------|-------|-------|
| **A. Compression predictor** | Three estimators + an in-memory online-learning blend (entropy + content-type + pattern + 30-day time-decayed history), plus Magika (label + confidence) and content-aware algo/level selection | `pkg/staging/compression_predictor.go`, `pkg/chunking/compression_estimator.go`, `pkg/compression/content_aware.go`, `pkg/detection/magika.go` |
| **B. Chunk-size optimizer** | Genuine multi-constraint optimization (memory/cost/perf) + adaptive shard count (#106) | `pkg/chunking/calculator.go`, `pkg/pipeline/adaptive_shard_calculator.go` |
| **C. Worker-pool scaler** | Runtime-adaptive throughput-feedback scaling with plateau detection (not fixed CPU×2) | `pkg/pipeline/adaptive_worker_pool.go` |
| **D. Cost forecaster** | Linear regression w/ R²/MAE/confidence intervals + burn-rate analytics. (Note: the advertised "4 models" are really 1 — exponential & moving-average are `TODO` stubs that fall back to linear.) | `pkg/aws/cost/forecasting.go` |
| **E. Failure predictor** | Rich **reactive** handling (error classification, retry/backoff, multiregion failover, BBR congestion) + an in-memory predictive scaffold; **no predictor, no persisted failure history** | `pkg/pipeline/error_classifier.go`, `pkg/multiregion/`, `pkg/aws/s3/predictive_adaptation_engine.go` |

So an ML model would **replace heuristics inside existing engines**, not add capability from scratch. That's a much smaller conceptual leap than the issue implies — but also means the marginal value is "better decisions than the current heuristics," which needs measurement to justify.

### The actual blocker: no training corpus

The vision's Phase 1 ("learn from historical data") assumes a telemetry substrate. Reality:
- **Cost data IS durable and production-wired** — the #246 `CostRecord` ledger persists timestamped per-upload cost/size/region/storage-class to `~/.cargoship/budgets.json` and/or S3. **Model D could train on this today.**
- **Per-upload manifests** (written to S3 every upload) durably capture the *actual outcomes* Models A/B would need — real compression ratio achieved, sizes, shard/chunk counts, timing. **But** each manifest is a standalone restore artifact: there's no index, aggregation, or "list manifests for learning" — no queryable corpus.
- **All runtime performance/failure signals are ephemeral** — Prometheus metrics are scrape-only; the staging `CompressionHistory` and `predictive_adaptation_engine` "learning" reset every process. Model C and E have **no persisted training data at all.**

**Verdict:** ~half the raw ingredients are already written to durable storage; none of it is joined into a training corpus. The honest Phase 1 is *not* "train models" — it's **build the telemetry aggregation layer** (ingest manifests + ledger + a snapshot of runtime metrics into a queryable history). Without that, four of five models can't be trained regardless of algorithm choice.

### Recommended scoped options (S/M/L)

**Option S — Close the telemetry loop + finish the honest heuristics (recommended first step).**
No ML. (1) Persist a compact per-upload outcome record (join manifest outcomes + ledger + runtime throughput/failure into one appended history, reusing the #246 store pattern). (2) Persist the staging `CompressionHistory` that already exists in-memory so its online learning survives restarts. (3) Implement the two TODO forecasting models (exp smoothing, weighted MA) so the "ensemble" is real. Outcome: a real training corpus exists, existing adaptive components actually improve across runs, and we can *measure* whether heuristics are even leaving value on the table — the prerequisite for justifying ML. Effort: **M** despite the "S" payoff framing; low risk, all additive.

**Option M — One data-driven model where the data is real: cost.**
On top of S, replace `forecasting.go`'s linear regression with a proper model (the ledger is genuine timestamped training data). Ship `cargoship cost forecast` improvements + a confidence-scored `estimate`. This is the *only* model with a production-wired dataset today, keeps everything offline/local (no telemetry-sharing/privacy surface), and delivers the vision's "95% cost-estimate accuracy" goal concretely. Effort: **M**.

**Option L — The full 5-model ensemble as specified.**
Federated learning, RL chunk optimizer, LSTM worker scaler, cross-user telemetry, differential privacy. This is a multi-quarter research program with a privacy/consent/telemetry-sharing surface, model-registry infra, and a CLI-size constraint (<10MB models). Only sensible *after* S proves the heuristics underperform and the corpus exists. Effort: **XL**; I'd not commit to it without the S/M evidence.

### How #137 and #140 fold in
- **#140** (adaptive-staging enhancements) = the `PerformanceAnalyzer` intelligence fields (`trendAnalyzer`/`patternDetector`/`anomalyDetector`/`predictionModel`) are stubbed `// TODO` in `pkg/aws/s3/adaptive_staging.go`. Those are exactly Option S's "persist + trend" work applied to staging. **#140 is a subset of Option S.**
- **#137** (ML shard routing) = Model B's optimizer with a learned policy. It's a subset of Option L and can't be justified until the corpus (Option S) shows the current multi-constraint calculator is suboptimal.

Suggest treating **#167 as the parent**, with **#140 ⊂ Option S** and **#137 ⊂ Option L**.

### Open product questions (the actual decision)
1. **Is offline/local-only acceptable, or is cross-user federated learning a goal?** This is the biggest fork — it determines whether the privacy/consent/telemetry-sharing machinery in the vision is in scope at all. (Option S/M are fully local; L is where sharing enters.)
2. **What's the success bar?** The vision cites 20% faster / 30% cheaper vs *naive* settings — but the baseline is already adaptive, not naive. Is "beat the current heuristics by X%" the real target, and is that measured before committing to ML?
3. **Ship-as-default or opt-in `--ai-optimize`?** Note the adaptive/staging transporters are currently opt-in, so even today's learning rarely runs on the default path.

My recommendation: **approve Option S now** (unlocks measurement + makes existing adaptivity durable, low risk), decide M after seeing the corpus, and gate L on evidence. Happy to scope Option S into concrete issues if that's the call.

**Comment by scttfrdmn** [comment]
### Option S scoped into issues

- **#261 — persist a per-upload telemetry/outcome record** (the training corpus; the keystone). No ML — joins the manifest outcomes + cost ledger + runtime counters into a durable, opt-in, append-only history. Unblocks everything else.
- **#262 — persist the staging CompressionRatioPredictor's learned history** (small; the online learner already exists in-memory, just make it survive restarts).
- **#263 — implement the stubbed exponential & moving-average cost forecasts** (small; finishes the honestly-fake "4-model" ensemble in `forecasting.go`).
- **#140** — the adaptive-staging `PerformanceAnalyzer` (trend/anomaly) layer; consumes #261's corpus. Sequence after #261.

Suggested order: **#263** and **#262** are independent quick wins; **#261** is the substrate; **#140** follows #261. After these land we can *measure* whether the current heuristics underperform — which is the evidence gate for Option M (data-driven cost model) and Option L (the full ensemble / #137). Leaving this parent open and `needs-decision` for the M/L call.

**Comment by scttfrdmn** [comment]
**Decision (2026-07-23): defer the ML options (M/L) for now.**

Option S — the measurement-first, no-ML substrate — is fully landed and released in **v0.15.0**:
- #263 real exponential & moving-average cost forecasts
- #262 persisted staging compression-ratio history
- #261 per-upload outcome corpus (the training data) + `cargoship cost history`
- #140 staging trend/anomaly analysis over that corpus (the first consumer)

That gives us the instrumentation to *measure* whether the current deterministic heuristics actually leave value on the table. Until the corpus shows they underperform, **Option M (data-driven cost model) and Option L (full ensemble / ML shard routing, #137) stay deferred** — no ML work is scheduled.

Leaving this issue open as the tracking home for that future decision; revisit once there's enough real-upload history to analyze.

**Comment by scttfrdmn** [comment]
Closing as premature.

Both ML tracks (this and #137) have been carried as deferred for several releases without being started, and the work the project actually needs is in a different direction: correctness, CLI contract fidelity (#316), and evidence accumulation. An ML optimizer layered on a pipeline whose `--compression-level` and `--shard-strategy` flags don't yet reach the pipeline would be optimizing controls that don't work.

There's also a sequencing argument. ML-driven optimization needs a trustworthy measurement substrate to learn from; the Option S measurement work (#261/#262/#263) laid groundwork but production-scale evidence is still thin. Deciding what to optimize before we can measure reliably is how you get a model that confidently tunes the wrong thing.

Not a rejection of the idea — an acknowledgement that keeping it open as a permanent "deferred" entry misrepresents the roadmap. Reopen when the measurement substrate and the CLI contract are solid.
